### ISSUE TITLE
[BUG] uv produces unusable requirements when pyproject.toml declares a custom index

### ISSUE THREAD

**Issue by nick-mal** [mlflow/mlflow#22881]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Databricks

### MLflow version

Client: 3.11.1
Tracking server: 3.11.1

### System information

OS Platform and Distribution: Databricks Runtime / Linux (Ubuntu)
Python version: 3.12.3

### Describe the problem

When `mlflow.flavor.log_model` is called in a project that has a pyproject.toml and uv.lock (either auto-detected or provided via the uv_project_path parameter), and the project declares a custom index, `uv export` produces a requirements list that omits the index information. 

The resulting requirements file is not installable downstream(for example in model serving), since the packages sourced from the custom index cannot be resolved.

While this appears to be intended behavior(see [issue in uv repo](https://github.com/astral-sh/uv/issues/13572)) of uv export itself, it effectively prevents using uv as a replacement for the existing dependency-inference path whenever a custom index is involved.

### Tracking information

MLflow version: 3.11.1
Python: 3.12.3

### Code to reproduce issue

pyproject.toml

```toml
[project]
name = "mlflow-uv-index-repro"
version = "0.1.0"
requires-python = ">=3.12"
dependencies = [
    "scikit-learn>=1.3",
    "my-internal-utils>=0.2",   # lives on a private index
]

[[tool.uv.index]]
name = "internal"
url = "https://pypi.internal.example.com/simple"
default = true

[tool.uv.sources]
my-internal-utils = { index = "internal" }
```

failure_demonstation.py
```python
import mlflow
from sklearn.linear_model import LinearRegression
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=50, n_features=4)
model = LinearRegression().fit(X, y)

# uv auto-detection kicks in because uv.lock + pyproject.toml exist in CWD
with mlflow.start_run() as run:
    info = mlflow.sklearn.log_model(model, name="model")
```

### Stack trace

An attempt to pip install this requirements would cause:

```
ERROR: Could not find a version that satisfies the requirement my-internal-utils==0.2.7
ERROR: No matching distribution found for my-internal-utils==0.2.7
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

**Comment by shagunt** [comment]
I see no assignments on this issue, i would like to work on it. How can i assign it to my self.

**Comment by Snehadas2005** [comment]
Hi @nick-mal, I can reproduce the issue where` uv export` omits custom indices, making requirements unusable for serving.

**Proposed Approach:**

- I'll investigate `mlflow/utils/environment.py` (or the relevant model logging utility) to identify where `uv export` is triggered.
- Since `uv export` intentionally omits indices, I'll implement a logic to parse the `[[tool.uv.index]]` blocks directly from `pyproject.toml` using `tomli` (or `tomllib` for Python 3.11+).
- I will then programmatically prepend the correct `--index-url` and `--extra-index-url` flags to the generated `requirements.txt` before it's packaged with the model.

I had worked on similar dependency management tasks and would love to fix this. Could you please assign this to me?

**Comment by serena-ruan** [comment]
@nick-mal Could you try if https://github.com/mlflow/mlflow/pull/22921 fixes your issue?

**Comment by nick-mal** [comment]
@serena-ruan 

Checked on my case, looks good to me. I can see index now in generated requirements, thanks.
